pkg/api/steve/collections/rbac.go

```diff
@@ -0,0 +1,315 @@
+		clusterRole, err = h.clusterRolesCache.Get(rb.RoleRef.Name)
+		if err != nil {
+			return nil, err
+		}
+		rules = clusterRole.Rules
+	} else {
+		role, err = h.rolesCache.Get(rb.GetNamespace(), rb.RoleRef.Name)
+		if err != nil {
+			return nil, err
+		}
+		rules = role.Rules
+	}
+	collectionRules, err := GetCollectionRules(h.apis, rules)
+	if err != nil {
+		if errors.Is(err, ErrNotReady) {
+			logrus.Debugf("not finished setting up project roles, retrying")
+			h.roleBindings.EnqueueAfter(rb.Namespace, rb.Name, enqueueAfterPeriod)
+			return nil, nil
+		}
+		return nil, err
+	}
+	if len(collectionRules) == 0 {
+		return rb, nil
+	}
+
+	subjects := rb.Subjects
+	roleRef := rb.RoleRef
+	roleRef.Name += suffix
+
+	collectionRoleBindingName := pkgrbac.NameForRoleBinding(namespace.Name, roleRef, subjects[0])
+	collectionRoleBinding, err := h.roleBindingsCache.Get(namespace.Name, collectionRoleBindingName)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return rb, nil
+	}
+	if collectionRoleBinding == nil {
+		collectionRoleBinding, err = h.createRoleBinding(collectionRoleBindingName, namespace.Name, roleRef, subjects, rb)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if collectionRoleBinding.RoleRef.Kind == "ClusterRole" {
+		collectionClusterRole, err := h.clusterRolesCache.Get(collectionRoleBinding.RoleRef.Name)
+		if err != nil && !apierrors.IsNotFound(err) {
+			return nil, err
+		}
+		if collectionClusterRole == nil {
+			err = h.createClusterRole(collectionRoleBinding.RoleRef.Name, collectionRules, clusterRole)
+			if err != nil {
+				return nil, err
+			}
+			return rb, nil
+		}
+		if _, ok := collectionClusterRole.GetAnnotations()[clusterRoleOwnerAnnotation]; ok { // owned by roletemplate controllers, don't update it
+			return rb, nil
+		}
+		err = h.compareAndUpdateClusterRole(collectionClusterRole, collectionRules, clusterRole)
+		if err != nil {
+			return nil, err
+		}
+		return rb, nil
+	}
+	collectionRole, err := h.rolesCache.Get(namespace.Name, collectionRoleBinding.RoleRef.Name)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return nil, err
+	}
+	if collectionRole == nil {
+		err = h.createRole(collectionRoleBinding.RoleRef.Name, namespace.Name, collectionRules, role)
+		if err != nil {
+			return nil, err
+		}
+		return rb, nil
+	}
+	err = h.compareAndUpdateRole(collectionRole, collectionRules, role)
+	if err != nil {
+		return nil, err
+	}
+
+	return rb, nil
+}
+
+func GetCollectionRules(apis APIResourceWatcher, rules []rbacv1.PolicyRule) ([]rbacv1.PolicyRule, error) {
+	collectionRules := []rbacv1.PolicyRule{}
+	validRules := []rbacv1.PolicyRule{}
+	for _, rule := range rules {
+		if len(rule.ResourceNames) != 0 {
+			continue
+		}
+		if len(rule.NonResourceURLs) != 0 {
+			continue
+		}
+		for _, verb := range rule.Verbs {
+			if verb == "*" || verb == "list" {
+				validRules = append(validRules, rule)
+			}
+		}
+	}
+	if len(validRules) == 0 {
+		return collectionRules, nil
+	}
+	if len(apis.List()) == 0 {
```

Comment:
I'm calling it here for two reasons - I'm calling GetCollectionRules from two places, so if this check is here I only have to have it in one place instead of two, and for the reason given above where it might not be necessary to check it at all and delaying doing so gets more objects through the work queue faster.

Reply:
Resolving, approach changed, file no longer exists.